packages/@lwc/engine-core/src/framework/wiring/types.ts

```diff
@@ -21,6 +21,7 @@ export interface WireAdapter {
 
 export interface WireAdapterConstructor {
     new (callback: DataCallback): WireAdapter;
+    new (callback: DataCallback, sourceContext: { tagName: string }): WireAdapter;
```

Comment:
See https://github.com/salesforce/lwc/pull/3487#issuecomment-1536826881 for perf comparison between passing an object vs. passing a string.

Reply:
Is there any specific reason we are overloading the `constructor` instead of updating the original method signature? Is there a case where we don't pass any context to the constructor?

Reply:
We don't have such a case.

Originally we were hoping to implement this in a "non-breaking way", and replacing the original method signature would've been a breaking type change on existing WireAdapters written in Typescript.

That said, we can accomplish that by updating the original method with the new parameter as optional and get rid of the overload.
```javascript
new (callback: DataCallback, sourceContext?: { tagName: string }): WireAdapter;
```

Reply:
I might be missing something, but adding a new parameter to a function shouldn't introduce a breaking change with TypeScript.

Reply:
Maybe it's our configuration or something to do with the rollup/plugin-typescript implementation then? On a [Stackblitz](https://stackblitz.com/edit/typescript-1g2rfh?file=index.ts) using bare Typescript, we get Typescript errors but the page still loads fine. Within LWC though, the build fails.

We unfortunately don't actually have any implementations of WireAdapterConstructor in our code, but if I add a new parameter to the [construct signature](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/base-lightning-element.ts#L132) of `LightningElementConstructor`: 
```javascript
export interface LightningElementConstructor {
    // new (): LightningElement;
    new (newParameter: string): LightningElement;
```
Our build breaks with the following error:
```
$ rollup --config scripts/rollup.config.js
       
       /Users/jye/dev/lwc/packages/@lwc/engine-core/src/index.ts → engine-core.js, engine-core.cjs.js...
       [!] (plugin typescript) RollupError: @rollup/plugin-typescript TS2554: Expected 1 arguments, but got 0.
       src/framework/invoker.ts (52:24)
       
       52         const result = new Ctor();
                                 ~~~~~~~~~~
       
         src/framework/base-lightning-element.ts:132:10
           132     new (newParameter: string): LightningElement;
                        ~~~~~~~~~~~~~~~~~~~~
           An argument for 'newParameter' was not provided.
```
where `Ctor: LightningElementConstructor`.

I've also seen the following error.
```
TS2322: Type '(value: any) => void' provides no match for the signature 'new (value: string, additionalValue: string)
```

I could also be missing something though, so let me know if we're talking about different things.

Reply:
>Expected 2 arguments, but got 1. when invoking new

Unfortunately, `Skilling-and-Enablement/ui-external-enablement`'s Storybook has a system of mocking wires for their features and invoking their wire adapter constructors, so 1 will be an issue for them. Otherwise, you're right, I couldn't find anyone else instantiating LWC wire adapters.

>Type '(value: any) => void' provides no match for the signature 'new (value: string, additionalValue: string): ConstructedObject'

I originally thought we had to worry about 2. because one of our biggest downstream users (Luvio) defines their adapters in Typescript with that function syntax. But I just checked their exact syntax, and they actually use:
```
return constructor as unknown as WireAdapterConstructor
```
which bypasses the whole check, so we don't need to worry about them.

>IMO it is preferable to tidy up our TypeScript definitions at this stage at the expense of potentially introducing a breaking change

Didn't we want to backport this into 244/patch? If that was the goal, no-breaking-changes would be preferred right?

---
In any case, making a single constructor with the new parameter as optional cleans up the overloaded constructors and resolves the errors so it should also be backport-able. Are you good with that approach?
```javascript
    new (callback: DataCallback, sourceContext?: { tagName: string }): WireAdapter;
```

Reply:
> Didn't we want to backport this into 244/patch? If that was the goal, no-breaking-changes would be preferred right?

Correct, the goal here is to backport this to `244/patch`. I am fine with this approach. 

My recommendation would be at least to get the types right in `main`. That said I don't have a strong opinion. Your call @jye-sf.

Reply:
>My recommendation would be at least to get the types right in main.

And just to clarify, your opinion is that `sourceContext` should be a non-optional parameter in `main` and downstream teams should update their implementations accordingly when they upgrade, right?

I think that's fair to do in `main`.